Sources/ValidatingProperty.swift

```diff
@@ -0,0 +1,307 @@
+	///   - validator: The closure to invoke for any proposed value to `self`.
+	public convenience init<U, E: Swift.Error>(
+		_ inner: MutableProperty<Value>,
+		with other: MutableValidatingProperty<U, E>,
+		_ validator: @escaping (Value, U) -> ValidatorOutput<Value, ValidationError>
+	) {
+		self.init(inner, with: other, validator)
+	}
+
+	/// Create a `MutableValidatingProperty` that validates mutations before
+	/// committing them.
+	///
+	/// The proposed value is only committed when `success` is returned by the
+	/// `validator` closure.
+	///
+	/// - parameters:
+	///   - initial: The initial value of the property. It is not required to
+	///              pass the validation as specified by `validator`.
+	///   - other: The property that `validator` depends on.
+	///   - validator: The closure to invoke for any proposed value to `self`.
+	public convenience init<U, E: Swift.Error>(
+		_ initial: Value,
+		with other: MutableValidatingProperty<U, E>,
+		_ validator: @escaping (Value, U) -> ValidatorOutput<Value, ValidationError>
+	) {
+		let otherValidations = other.result
+
+		self.init(initial) { input in
+			let otherValue: U
+
+			switch otherValidations.value {
+			case let .success(value):
+				otherValue = value
+			case let .substitution(_, value, _):
+				otherValue = value
+			case let .failure(value, _):
+				otherValue = value
+			}
+
+			return validator(input, otherValue)
+		}
+
+		otherValidations.signal
+			.take(during: lifetime)
+			.observeValues { [weak self] _ in
+			if let s = self {
+				switch s.result.value {
+				case let .failure(failedValue, _):
+					s.value = failedValue
+				case let .substitution(substitutedValue, _, _):
+					s.value = substitutedValue
+				case let .success(value):
+					s.value = value
+				}
+			}
+		}
+	}
+}
+
+/// Represents a decision of a validator of a validating property made on a
+/// proposed value.
+public enum ValidatorOutput<Value, Error: Swift.Error> {
+	/// The value is valid.
+	case success
+
+	/// The value is invalid, but the validator can provide a substitution derived
+	/// from the invalid value that would be considered valid.
+	case substitution(Value, Error?)
+
+	/// The value is invalid.
+	case failure(Error)
+}
+
+/// Represents the result of the validation performed by a validating property.
+public enum ValidationResult<Value, Error: Swift.Error> {
+	/// The value passed the validation.
+	case success(Value)
+
+	/// The value was a substituted value due to a failed validation.
+	case substitution(substituted: Value, proposed: Value, Error?)
+
+	/// The value failed the validation.
+	case failure(proposed: Value, Error)
+
+	/// Whether the validation was failed.
+	public var isFailure: Bool {
+		if case .failure = self {
+			return true
+		} else {
+			return false
+		}
+	}
+
+	/// Extract the value if the validation is passed, or a substituted value
+	/// for failure is provided.
+	public var value: Value? {
```

Comment:
Do you have a use case in mind for this?

I think this maybe shouldn't return an `Optional`. You could still check `isFailure` if you didn't want failed values. This would simplify the code that handles dependencies, e.g.